client/post-editor/editor-html-toolbar/index.jsx

```diff
@@ -23,12 +34,76 @@ export class EditorHtmlToolbar extends Component {
 	};
 
 	state = {
+		isPinned: false,
+		isScrollable: false,
+		isScrolledFull: false,
 		openTags: [],
 		selectedText: '',
 		showImageDialog: false,
 		showLinkDialog: false,
 	};
 
+	componentDidMount() {
+		this.pinToolbarOnScroll = throttle( this.pinToolbarOnScroll, 50 );
+		this.disablePinOnSmallScreens = throttle( this.disablePinOnSmallScreens, 400 );
+		this.toggleToolbarScrollableOnResize = throttle( this.toggleToolbarScrollableOnResize, 200 );
+		this.hideToolbarFadeOnFullScroll = throttle( this.hideToolbarFadeOnFullScroll, 200 );
+
+		window.addEventListener( 'scroll', this.pinToolbarOnScroll );
+		window.addEventListener( 'resize', this.disablePinOnSmallScreens );
+		window.addEventListener( 'resize', this.toggleToolbarScrollableOnResize );
+		this.buttons.addEventListener( 'scroll', this.hideToolbarFadeOnFullScroll );
+
+		this.toggleToolbarScrollableOnResize();
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'scroll', this.pinToolbarOnScroll );
+		window.removeEventListener( 'resize', this.disablePinOnSmallScreens );
+		window.removeEventListener( 'resize', this.toggleToolbarScrollableOnResize );
+		this.buttons.removeEventListener( 'scroll', this.hideToolbarFadeOnFullScroll );
+	}
+
+	bindButtonsRef = div => {
+		this.buttons = div;
+	}
+
+	pinToolbarOnScroll = () => {
+		if ( isWithinBreakpoint( '<660px' ) ) {
+			return;
+		}
+
+		const { offsetTop } = this.props.content;
+		const { isPinned } = this.state;
+
+		if ( isPinned && window.pageYOffset < offsetTop - TOOLBAR_HEIGHT ) {
+			this.setState( { isPinned: false } );
+		} else if ( ! isPinned && window.pageYOffset > offsetTop - TOOLBAR_HEIGHT ) {
+			this.setState( { isPinned: true } );
+		}
+	}
+
+	disablePinOnSmallScreens = () => {
+		if ( isWithinBreakpoint( '<660px' ) ) {
+			this.setState( { isPinned: false } );
+		}
+	}
+
+	toggleToolbarScrollableOnResize = () => {
+		this.setState( {
```

Comment:
Can we also consider merging with `disablePinOnSmallScreens` into a single `resize` handler?